### DISCUSSION TITLE
raspiCam

### DISCUSSION THREAD

**Discussion by bberg1** [Lillifee/raspiCam#72] [General]
Thank you, I finally can stream live video :-)

Is there a way to stop/disable and free up all of the assets?  Once I stop server.js I can no longer get to libcamera-vid.

Regards,
-Bill

**Comment by Lillifee** [comment]
Hey @bberg1 

When the server shuts down, it should also terminate all child processes, including the libcamera. I frequently start and stop the server during development, and the libcamera is closed accordingly. 

Do you manually run it, or is it configured as a service? Also, are there any errors in the application's output upon exiting?
That also implies that you can start the server just once, as the camera remains in use the second time? 
Could you also verify or compare the processes before and after stopping the server using ```ps aux``` or ```top```?

**Comment by bberg1** [comment]
Hey @bberg1 <https://github.com/bberg1> 
Hey Lillifee,
*Thank you for responding*

Just so you know, I’m very new to the Pi and Linux.  A bunch of years ago I purchased a 3B+ so I could set up a camera on my 3D printer and watch it remotely so I didn’t have run up and down the stairs a hundred times to see if needed attention.  I believe this was installed with NOOBS back then with the Stretch OS version.  I found advice online and it has worked well except it occasionally would lock up and needed to be rebooted.  Have no idea why.  

Just to be ahead of a possible failure and my interest in learning more about linux I purchased a couple more of 3B+’s.  Was hoping that I could just insert the old SD and it would take off.  It would even boot!  Guess the hardware on 3B+ has changed over the years and it’s not backwards compatible?

Used rpi-imager to install Bulleseye full x(64) and copied the old software/setup – but no go.  Seems stuff has been deprecated and raspicam no longer works L

Did a lot of web searching and trying options to live stream on the web.  Yours was the FIRST that seemed to work.

Enough background.

I’ve followed your guidelines multiple times.  Have run it standalone and also as a service.  Works great.  And it would certainly meet my first objective for using it to monitor the 3D printer.

But wanting the explore/learn things further I kept moving.  This is when I first noticed that Libcamera-vid would not work after I ended it raspiCam/node server.js with cntrl-z..  Maybe I’m not really stopping it correctlyng?

Anyway, I’ll try to respond to your questions.

Fresh boot – server.js and raspiCam previously installed, not running as a service
RealVNC
Terminal window
ps aux – saved as PS-1
libcamera-vid –t0 - works
            close window
ps aux – saved as PS-2
cd raspiCam - 
node server.js – save screen PS-3
ctrl-z to exit
ps aux -  saved as PS-4
libcamera-vid – window flashes and closes
saved screen to PS-5
ps aux – saved to PS-6

Files attached.

Thank you for your help,

-B

When the server shuts down, it should also terminate all child processes, including the libcamera. I frequently start and stop the server during development, and the libcamera is closed accordingly.

Do you manually run it, or is it configured as a service? Also, are there any errors in the application's output upon exiting?
That also implies that you can start the server just once, as the camera remains in use the second time?
Could you also verify or compare the processes before and after stopping the server using ps aux or top?

**Comment by bberg1** [comment]
Lillifee,

I think this is probably a very newbie issue J

I did some research and found that the appropriate way(?) is to use cntrl-C to cancel.  After that everything seems to work as suspected.  I can run server.js and libcamera-vid as much as I want as long as I cancel with cntrl-C.

 

I’m learning

 

Regards,

-B

From: Bill ***@***.*** 
Sent: Friday, March 22, 2024 5:23 PM
To: 'Lillifee/raspiCam'; 'Lillifee/raspiCam'
Cc: 'Mention'
Subject: RE: [Lillifee/raspiCam] raspiCam (Discussion #72)

 

Hey @bberg1 <https://github.com/bberg1> 
Hey Lillifee,
*Thank you for responding*

Just so you know, I’m very new to the Pi and Linux.  A bunch of years ago I purchased a 3B+ so I could set up a camera on my 3D printer and watch it remotely so I didn’t have run up and down the stairs a hundred times to see if needed attention.  I believe this was installed with NOOBS back then with the Stretch OS version.  I found advice online and it has worked well except it occasionally would lock up and needed to be rebooted.  Have no idea why.  

Just to be ahead of a possible failure and my interest in learning more about linux I purchased a couple more of 3B+’s.  Was hoping that I could just insert the old SD and it would take off.  It would even boot!  Guess the hardware on 3B+ has changed over the years and it’s not backwards compatible?

Used rpi-imager to install Bulleseye full x(64) and copied the old software/setup – but no go.  Seems stuff has been deprecated and raspicam no longer works L

Did a lot of web searching and trying options to live stream on the web.  Yours was the FIRST that seemed to work.

Enough background.

I’ve followed your guidelines multiple times.  Have run it standalone and also as a service.  Works great.  And it would certainly meet my first objective for using it to monitor the 3D printer.

But wanting the explore/learn things further I kept moving.  This is when I first noticed that Libcamera-vid would not work after I ended it raspiCam/node server.js with cntrl-z..  Maybe I’m not really stopping it correctlyng?

Anyway, I’ll try to respond to your questions.

Fresh boot – server.js and raspiCam previously installed, not running as a service
RealVNC
Terminal window
ps aux – saved as PS-1
libcamera-vid –t0 - works
            close window
ps aux – saved as PS-2
cd raspiCam - 
node server.js – save screen PS-3
ctrl-z to exit
ps aux -  saved as PS-4
libcamera-vid – window flashes and closes
saved screen to PS-5
ps aux – saved to PS-6

Files attached.

Thank you for your help,

-B

When the server shuts down, it should also terminate all child processes, including the libcamera. I frequently start and stop the server during development, and the libcamera is closed accordingly.

Do you manually run it, or is it configured as a service? Also, are there any errors in the application's output upon exiting?
That also implies that you can start the server just once, as the camera remains in use the second time?
Could you also verify or compare the processes before and after stopping the server using ps aux or top?

**Comment by Lillifee** [comment]
Hey @bberg1,

It's a bit challenging to pinpoint what caused the issue with your old Pi. Raspberry Pi has transitioned from using raspi-still, raspi-vid, etc., to utilizing libcamera apps. 

Nevertheless, it's fantastic that you were able to diagnose the problem on your own! Just to clarify, using ctrl z only suspends the process, leaving the server running in the background and blocking other camera usage. To fully stop it, you'd need to use ctrl c.

**Comment by bberg1** [comment]
Hey @bberg1 <https://github.com/bberg1> ,

It's a bit challenging to pinpoint what caused the issue with your old Pi. Raspberry Pi has transitioned from using raspi-still, raspi-vid, etc., to utilizing libcamera apps.

Understood.  But I would have expected it to at least boot and just not run the script.

NBD.  It still runs and now I have a backup.

Thank you,

-B

Nevertheless, it's fantastic that you were able to diagnose the problem on your own! Just to clarify, using ctrl z only suspends the process, leaving the server running in the background and blocking other camera usage. To fully stop it, you'd need to use ctrl c.

**Comment by bberg1** [comment]
Lillifee

If I accidently use the ctrl-z to suspend the service, how can I restart it.  Just using the node command doesn’t work.  And if I do reboot that doesn’t work either.  Have disconnect/reconnect power.

 

NBD, just wondering.

-B

**Comment by Lillifee** [comment]
You can use the command ```fg``` to bring a process to the foreground and ```bg``` to run it in the background. However, upon restarting your Pi, all processes should be terminated, and no process should remain suspended. These are linux command line features and not related to raspicam...

https://prateeksrivastav598.medium.com/understanding-ctrl-z-and-ctrl-c-in-the-linux-terminal-c28bd1e00dbe

**Comment by bberg1** [comment]
Thank you, again!

-B

**Comment by bberg1** [comment]
Another question…

Can you point me to a site that will show how to set up over the internet connection?

I already use a DDNS service that works well for the outside cameras, but when I try to connect the Pi via http I get the msg ‘refused to connect’.

So I’m guessing additional configuration is required.

 

Thank you,

-B

**Comment by bberg1** [comment]
Lillifee,

I think I have working now.

 

Thks,

-B

 

From: Bill ***@***.*** 
Sent: Sunday, March 24, 2024 7:27 PM
To: 'Lillifee/raspiCam'; 'Lillifee/raspiCam'
Cc: 'Mention'
Subject: RE: [Lillifee/raspiCam] raspiCam (Discussion #72)

 

Another question…

Can you point me to a site that will show how to set up over the internet connection?

I already use a DDNS service that works well for the outside cameras, but when I try to connect the Pi via http I get the msg ‘refused to connect’.

So I’m guessing additional configuration is required.

 

Thank you,

-B

**Comment by Lillifee** [comment]
Hey @bberg1,

It's great that you found a solution. However, I need to point out that Raspicam isn't designed to function as a server for a public website. It lacks password protection and similar security features. Additionally, it's challenging to keep up with updates and maintenance to address potential vulnerabilities of node. 

My suggestion is to run it on your local network and set up VPN access to your router. Many routers offer straightforward solutions, eliminating the need to expose it to the public web. This approach enhances security and ensures safer operation.

**Comment by bberg1** [comment]
Never simple J

I have the Pi set with a static IP

I use a DDNS

RaspiCam is started up with a port, –p option

The router is set up for port forwarding to point to RaspiCam

 

And from here folks could theoretically get to my network assets?

 

-B

**Comment by Lillifee** [comment]
Hey @bberg1,

Just a heads-up: the camera is online, which means someone could potentially access it. While it might not seem like a big deal to have the 3D printer visible, attackers could access the web interface to mess with the settings. And if they find a vulnerability, like in Node.js, they might be able to sneak into your Raspberry Pi and from there, your network.

Remember, no software is foolproof. New vulnerabilities pop up all the time, and without regular updates, things could get risky. Additionaly there is no authentication etc.

For example, I'm using Node.js for the server, and new vulnerabilities published every day. 
https://www.cvedetails.com/vulnerability-list/vendor_id-12113/Nodejs.html

However, this applies to anything exposed to the web. But for instance, a router with a VPN should be relatively safe, as it operates using highly secure protocols with strong encryption. The risk is significantly reduced in such cases.

**Comment by bberg1** [comment]
Hi Lillifee,

If you are getting annoyed with all of my questions don’t be hesitant to say so.  Don’t want to be a nuisance.

 

How can I change additional parameters for the libcamera-vid process?  Like maybe –hflip or others.

 

I’ve looked at the server.js file but nothing jumped out at me.

 

Regards,

-B

**Reply by Lillifee** [reply]
Hey @bberg1 

those are camera settings. You can find the hflip, vflip, rotation etc. in the Camera -> Other settings.

**Comment by bberg1** [comment]
Apologize for noob questions – where is ‘Camera->Other’ located?

 

Only thing I know how to do is add commands/parameters to Libcamera like ‘libcamer-vid -t0 –hflip

 

-B

**Comment by Lillifee** [comment]
Hey @bberg1 

oh i ment on the webinterface under the settings
![image](https://github.com/Lillifee/raspiCam/assets/23337626/8c007646-c1fc-41da-9123-93f39a58eee9)

**Comment by bberg1** [comment]
Perfect!

 

So… appears it only changes settings for the node-server.js  If I run libcamera from terminal window it’s back to default, whatever that is.

 

So how in the heck do changes made via an http streaming session update node-server.js

 

So many unknowns J

 

-B

**Comment by Lillifee** [comment]
Hey @bberg1 

if you want to use raspicam to find the best settings, you can adjust it in the web interface and copy the command from the terminal output to run libcamera with the same setting.

This is an old screenshot with raspivid, but you get the full libcamera command in the terminal as well.
![image](https://github.com/Lillifee/raspiCam/assets/23337626/a6e704ba-2de3-496a-8603-196f549692b3)

What do you want to archive by running the libcamera manually in the terminal?

**Comment by bberg1** [comment]
Lillifee,

I’m mostly just trying to educate myself, no interest in archiving anything.

 

You may recall that I use the Pi to monitor the 3D printer.  It was setup years ago I think on the stretch version.  It started to occasionally freeze and I had to reboot it.  I’ve since also connected it via Ethernet and so far it has worked good.  Maybe the wireless connection was burping or something that caused the freeze?  Or maybe it will occur again, but so far not.

 

Anyway, thinking maybe the old thing was going dead I purchased two more Pi 3B+ to have a backup and for education.  Hence, I’ve been bugging you J

 

Yours is the first option I D/L’d that worked, so that was progress and appreciated.

 

One thing I’ve noticed is that running the old version when I link in via http I can enlarge/reduce the image and use the mouse to move the image around.  This doesn’t seem to be the same with server.js.  NBD.  At least now I have a backup if needed.  And I can continue to experiment with the new units.

 

I’ve already purchased a breakout terminal block for the GPIO.  But no projects in the works.  I do recall using similar setups when working with interfacing MCU’s.  Will have to reeducate myself on that too.

 

Today I’m going to connect a Bluetooth speaker and set something to say hello or play a audio file when it boots.

 

Thks for your replies,

-B
